source/common/router/scoped_rds.cc

```diff
@@ -100,71 +108,182 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources,
     const std::string& version_info) {
-  std::vector<envoy::api::v2::ScopedRouteConfiguration> scoped_routes;
-  for (const auto& resource_any : resources) {
-    scoped_routes.emplace_back(MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(
-        resource_any, validation_visitor_));
+  bool any_applied = false;
+  std::vector<std::string> exception_msgs;
+  absl::flat_hash_set<std::string> unique_resource_names;
+  envoy::config::filter::network::http_connection_manager::v2::Rds rds;
+  rds.mutable_config_source()->MergeFrom(rds_config_source_);
+
+  // If new route config sources come after the factory_context_.initManager()'s initialize() been
+  // called, that initManager can't accept new targets. Instead we use a local override which will
+  // start new subscriptions but not wait on them to be ready.
+  std::unique_ptr<Init::ManagerImpl> overriding_init_manager;
+  if (factory_context_.initManager().state() == Init::Manager::State::Initialized) {
+    // Pause RDS to not send a burst of RDS requests until we start all the new subscriptions.
+    factory_context_.clusterManager().adsMux().pause(
+        Envoy::Config::TypeUrl::get().RouteConfiguration);
+    overriding_init_manager =
+        std::make_unique<Init::ManagerImpl>(fmt::format("SRDS {}:{}", name_, version_info));
   }
 
-  std::unordered_set<std::string> resource_names;
-  for (const auto& scoped_route : scoped_routes) {
-    if (!resource_names.insert(scoped_route.name()).second) {
-      throw EnvoyException(
-          fmt::format("duplicate scoped route configuration {} found", scoped_route.name()));
+  std::list<ScopedRouteInfoConstSharedPtr> to_be_deleted;
+  for (const auto& resource : added_resources) {
+    envoy::api::v2::ScopedRouteConfiguration scoped_route_config;
+    try {
+      scoped_route_config = MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(
+          resource.resource(), validation_visitor_);
+      MessageUtil::validate(scoped_route_config);
+      if (!unique_resource_names.insert(scoped_route_config.name()).second) {
+        throw EnvoyException(fmt::format("duplicate scoped route configuration '{}' found",
+                                         scoped_route_config.name()));
+      }
+      rds.set_route_config_name(scoped_route_config.route_configuration_name());
+      ScopedRouteInfoConstSharedPtr scoped_route_info = std::make_shared<ScopedRouteInfo>(
+          std::move(scoped_route_config),
+          route_config_provider_manager_.createRdsRouteConfigProvider(
+              rds, factory_context_, stat_prefix_,
+              (overriding_init_manager == nullptr ? factory_context_.initManager()
+                                                  : *overriding_init_manager)));
+      // Detect if there is key conflict between two scopes.
```

Comment:
This is subtle and worth expanding on in comments. What's the expected behavior from the operator?